### PR TITLE
DataFrame: Add metadata for supporting time comparison windows

### PR DESCRIPTION
**Pull request by dprokop**
This builds on top of exploration done in https://github.com/grafana/grafana/pull/71204 and https://github.com/grafana/grafana/pull/71129. This PR adds `timeRange` property to data frame metadata, to indicate what time range a query was executed with.

Spent some time talking with @leeoniya about this and we think the approach of providing not pre-shifted data frames from Scenes will be sensible going forward. The time comparison graphing support comes down to visualizations, and with this PR it is quite a minimal change to the panels in order to support rendering of compare series. The only thing required to do seems to be just shifting the compare series relative to reference (original time range) frame.

This PR assumes that queries run for a previous time period will be returned in data frames that follow the reference frame: `[referenceFrame, shiftedFrame1, shiftedFrame2,...]`. grafana/data exposes `shiftComparisonFramesTimestamps` utility function that is used by panels when they prepare data for rendering. This way, the changes required in panels are minimal.

@ryantxu  - I know you started exploring this territory- is this anyhow aligned with your approach?

### CODE REVIEW

**Comment by dprokop** [comment]
> Is it not better to shift time stamps outside the panels (so it's more compatible with different visualizations?)

I think both approaches have their pros and cons. But I do agree with @leeoniya  it should be up to the viz to decide what to do(if anything) with data from different time ranges

IMO the approach of pre-shifting in Scenes is driven by the constraints of uPlot-based visualizations (requires time stamps to be in the same time range in order to be properly aligned and graphed). And I'm not convinced we should build new functionalities with those limitations as drivers, but rather approach it from a more future-proof POV. 

Regardless of whether or not we pre-shift in Scenes, we will still land in a situation where some of the visualizations need adjusting (i.e. table - tho it should not be wildly used for comparison, or stat - if we want comparison in it, we would still have to calculate the comparison period in order to show it somehow in the stat panel). 

With this work, there are things that are not working tho. Those are for instance tooltips. But, with this approach, it will be trivial to display the correct time stamp for the comparison series in a tooltip, as we will just read it from the frame, without doing any calculations to un-shift the pre-shifted timestamps. However, this will follow the tooltip refactor (currently in progress), as the tooltip now reads the data from the aligned data frame, not from the original frames that came to the panel.

**Comment by leeoniya** [comment]
we also discussed whether time comparisons are simply a subset of general diffs. for example, with pyroscope flamegraph, wr have diffing of two different profile frames that are not necessarily offsrt by time but can be from different machines, or different git commits. we could have a code diffing view, or onion-skin image diffing ovetlays, or other forms of non-tine comparisons supported by different panels or UIs.

a lot of this can be done at the panel level with some "compare" mode and manual selection of frames to compare. so not pre-shifting or pre-processing the data ahead of the panel feels like the right way to go. we will just add metadata so that these compare modes can provide meaningful context, but retain original data.

**Comment by torkelo** [comment]
This require dataframe modifications inside every panel. I think it's much better if this is done by the query/data transform layer outside the panel. That way we can use this right away in a11y, without requiring latest Grafana. 

Then each panel can opt in to visualize time shifted data some custom way (like unshifting in the tooltip).

This seem like the worst of all options , requiring custom DataFrame manipulation in all panels handle this type of data.

Not sure I see the advantage to this compared to having the time shifting in a single place, and marking the frame with shift amount meta data

**Comment by torkelo** [comment]
> a lot of this can be done at the panel level with some "compare" mode and manual selection of frames to compare. so not pre-shifting or pre-processing the data ahead of the panel feels like the right way to go. we will just add metadata so that these compare modes can provide meaningful context, but retain original data.

Frame selection and compare mode feels like things that belong outside the visualizations.

The time window comparison and the query time shifting logic is outside. It knows how to shift data. So feels less complex to keep all the timeshift logic in one place to start with

File: packages/grafana-data/src/types/data.ts [hunk 1]
```diff
@@ -101,6 +101,11 @@ export interface QueryResultMeta {
   limit?: number; // used by log models and loki
   json?: boolean; // used to keep track of old json doc values
   instant?: boolean;
+
+  /**
+   * Indicates a time range that the data source was queried with.
+   */
+  timeRange?: TimeRange;
```
Comment (ryantxu) [inline packages/grafana-data/src/types/data.ts:108]:
```suggestion
  queryTimeRange?: AbsoluteTimeRange;
```

The TimeRange object is not great for serialization 😬 

Also I think we should make it clear that this with the query time range -- the data range is already in the data :)

File: packages/grafana-data/src/dataframe/index.ts [hunk 2]
```diff
@@ -7,5 +7,10 @@ export * from './dimensions';
 export * from './ArrayDataFrame';
 export * from './DataFrameJSON';
 export * from './frameComparisons';
-export { anySeriesWithTimeField, isTimeSeriesFrame, isTimeSeriesFrames } from './utils';
+export {
+  anySeriesWithTimeField,
+  isTimeSeriesFrame,
+  isTimeSeriesFrames,
+  shiftComparisonFramesTimestamps,
```
Comment (ryantxu) [inline packages/grafana-data/src/dataframe/index.ts:14]: does this need to be exported yet?  can it be an internal detail that we improve before making public?

File: packages/grafana-data/src/dataframe/utils.ts [hunk 3]
```diff
@@ -26,3 +26,34 @@ export function anySeriesWithTimeField(data: DataFrame[]) {
...
+ * of the frames so that they are graphable.
+ * This function assumes that the first frame contains the reference time range, while consecutive frames
+ * are time shifted.
+ */
+export function shiftComparisonFramesTimestamps(frames: DataFrame[]) {
+  // First frame that holds a time field that's used as comparison diffs reference
+  let timeComparisonReferenceFrame: DataFrame | undefined;
+
+  const referenceFrame = frames[0];
+  for (let field of referenceFrame.fields) {
```
Comment (ryantxu) [inline packages/grafana-data/src/dataframe/utils.ts:41]: IIUC, we only want to shift if:
1. all frames have a queryTimeRange
2. the values in the queryTimeRange are different (may not be the *same* instance)

File: packages/grafana-data/src/dataframe/utils.ts [hunk 4]
```diff
@@ -26,3 +26,34 @@ export function anySeriesWithTimeField(data: DataFrame[]) {
...
+      if (timeComparisonReferenceFrame && frame !== timeComparisonReferenceFrame) {
+        const diff = timeComparisonReferenceFrame.meta?.timeRange?.from.diff(frame.meta?.timeRange?.from!);
+        if (f.type === FieldType.time) {
+          f.values = f.values.map((v) => {
+            return v + (diff || 0);
+          });
+        }
+      }
+    });
+  });
```
Comment (torkelo) [inline packages/grafana-data/src/dataframe/utils.ts:58]: This function feels more fragile and error-prone, having to kind of reverse engineer the time shift amount (compared to timeshifting as part of a transform applied outside where we apply the same timeshift to the query range).

**Review by torkelo — COMMENTED** [review]
I don't understand the advantage of this approach compared and keeping the data transform outside like other transforms. https://github.com/grafana/grafana/pull/71129/ also addresses the naming of the timeshifted series which this PR does not. 

It feels a lot more fragile and complex to the timeshift logic spread out. Is not better to have this more high-level logic (data manipulation) outside the visualizations? 

by doing the transform outside like in the original POC this works with all visualizations without any modifications. Visualizations could still opt-in to customizing the handling (not sure I see any need for that right now). 

What am I missing that warrants this much more complex solution? Some future special visualization method for showing the original timestamps? Nothing preventing us from adding that later if we see a need for it (either by shifting timestamps back based on the meta data containing the shift amount, or by some other mechanism).

**Comment by dprokop** [comment]
> I don't understand the advantage of this approach compared and keeping the data transform outside like other transforms. #71129 also addresses the naming of the timeshifted series which this PR does not.
> 
> It feels a lot more fragile and complex to the timeshift logic spread out. Is not better to have this more high-level logic (data manipulation) outside the visualizations?
> 
> by doing the transform outside like in the original POC this works with all visualizations without any modifications. Visualizations could still opt-in to customizing the handling (not sure I see any need for that right now).
> 
> What am I missing that warrants this much more complex solution? Some future special visualization method for showing the original timestamps? Nothing preventing us from adding that later if we see a need for it (either by shifting timestamps back based on the meta data containing the shift amount, or by some other mechanism).

Indeed, the display name here is not implemented, but should be trivial given that display name calculation has access to all frames.


As I mentioned earlier, both approaches have pros and cons. But IMO the initial POC (https://github.com/grafana/grafana/pull/71129) suffers from the original sin - the data passed to the panel is not reflecting the truth about it. 

One can argue that we do a lot of processing prior to pushing the data to a panel (options, field config, transformations) but none of these manipulations affect the original data. This PR approach is conceptually similar to transformations-leave the original frames untouched and produce new ones when needed. So technically - to avoid changes in visualizations - we could introduce a shift transformation that makes the data passed to visualization pre-shifted. Tho it's still gonna cause issues i.e. when we want to render comparison series ts in a tooltip.

Problems I see with the initial POC that we gonna have to solve in the core soon are:
- What to do with transformations? I.e. join by time will produce incorrect results unless we un-shift the data prior to the transformation. This introduces plenty of complexity to the transformation implementation just for the sake of compensating for this inconvenience.
- Panel inspection will show incorrect data and will require un-shifting timestamps prior to rendering frames table.

To be clear - I do think the #71129 is a low-hanging fruit for us and a simple approach. But I also think it's built on a wrong assumption - current visualization limitations (primarily uPlot requirement for series alignment) that IMO should not be a driver for quite an important feature.

**Comment by torkelo** [comment]
> the data passed to the panel is not reflecting the truth about it.

It reflects what should be visualized (what the vis should care about) is that not enough?

**Comment by torkelo** [comment]
> What to do with transformations? I.e. join by time will produce incorrect results unless we un-shift the data prior to the transformation. This introduces plenty of complexity to the transformation implementation just for the sake of compensating for this inconvenience.

Not sure I understand this, won't this be a bigger issue when not shifting the time points? Is there a scenario when you want to join by time on two time-shifted series but not the original time?  Can't think of a real scenario here that makes any sense. 

> Panel inspection will show incorrect data and will require un-shifting timestamps prior to rendering the frames table.

I don't think this is a big issue, we don't need to unshift. this feature is for visual comparison. if you want to download raw data from a specific time range this is not the feature to do that. 

> But I also think it's built on a wrong assumption - current visualization limitations (primarily uPlot requirement for series alignment) that IMO should not be a driver for quite an important feature.

For me, series alignment is not the driver at all for me, but rather keeping visualizations as simple as possible and free from data transformations, to simplify design and reduce the responsibilities of different components.

**Comment by leeoniya** [comment]
> For me, series alignment is not the driver at all for me, but rather keeping visualizations as simple as possible and free from data transformations, to simplify design and reduce the responsibilities and interdependence of different components.

you can say the same about stacking, no? it's also shifting but on the y axis. why should the panels stack the data instead of a transform or some pre-processing ahead of the panel?

comparing over time does not necessarily mean the time  has to be shifted. even if this has to be true for TimeSeries (today) it is not true for other panel types. requiring panels to do this "reverse" transform to get back the original data is not something that any panels currently have to do and introducing this pattern is worse than allowing panels that understand shifted queries to implement the comparison as they see fit.

**Comment by torkelo** [comment]
With stacking I can see the value in not modifying the data outside the visualization as you might care about the unstacked value (say in the tooltip). In all the time range comparisons we have looked at none show the orignal time values (in a tooltip or raw data), so it does not seem relevant for this case of overlaying data from timeshifted range. 

So for graphs I don't see any need to do this data shifting in the visualization layer, Graphite has supported this time shift in the DB layer for over 13 years without any need to move it to vis layer. Kibana also does this timeshift in frontend in response processing layer (I think Elasticsearch can also do it in the DB layer).

If other visualization forms for comparing time series data arise we can just simply disable the time shifting for those scenarios. But right now the only thing we really need is comparing line graphs on the non shifted time axis. So for this scenario It just seems a lot less complex to do the timeshifting in the same place that does the query time range shifting. And have it be configured and controlled from there (we can easily add an option there to disable timestamp timeshifting). Here the comparison to stacking is also not a great match, as this feature is also a query execution option (time range), so it makes sense the timeshifting is in the query response processing (to keep this logic isolated to a single component, instead of in every graph vis).

**Comment by dprokop** [comment]
Closing as we decided to proceed with https://github.com/grafana/grafana/pull/71129

**Pull request closed without merging**